### PR TITLE
[improvement][client-java] Avoid too large memory preallocation for batch message.

### PR DESCRIPTION
**Pull request by tjiuming**
Fixes (https://github.com/apache/pulsar/issues/14943)

### Motivation

Allocate small memory for batch message first, and let it grow(with no memory resizes and memory copies).

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Review by codelipenghui — APPROVED** [review]

**Review by Jason918 — APPROVED** [review]
LGTM

**Comment by tjiuming** [comment]
@codelipenghui seems the PR is able to merge

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java [hunk 1]
```diff
@@ -69,8 +69,7 @@ public boolean add(MessageImpl<?> msg, SendCallback callback) {
                 messageMetadata.setSequenceId(msg.getSequenceId());
                 lowestSequenceId = Commands.initBatchMessageMetadata(messageMetadata, msg.getMessageBuilder());
                 this.firstCallback = callback;
-                batchedMessageMetadataAndPayload = PulsarByteBufAllocator.DEFAULT
-                        .buffer(Math.min(maxBatchSize, ClientCnx.getMaxMessageSize()));
+                batchedMessageMetadataAndPayload = PulsarByteBufAllocator.DEFAULT.compositeBuffer();
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java:72]: This approach could result in many buffer expansions when the average buffer size is larger than the "default" size of allocated buffers.

Comment (tjiuming) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java:72]: Buffer expansions will happens in some situations, but for `CompositeByteBuf`, it wouldn't takes too much costs.
`mem_copy` operations triggered when `CompositeByteBuf#consolidate0`, which means after write more than 1000000 bytes into buffer, `mem_copy` will be triggered first time by default. I believe it's acceptable in the `client side`

**Review by 315157973 — APPROVED** [review]

**Comment by tjiuming** [comment]
@codelipenghui Please help merge the PR

**Comment by Jason918** [comment]
> @codelipenghui Please help merge the PR

Please fix the CI failure

**Comment by AnonHxy** [comment]
@tjiuming  Please help fix the CI failure

**Comment by AnonHxy** [comment]
Please rebase the lastest master branch and push it force . `pulsarbot run-failure-checks` will not help now @tjiuming

**Review by AnonHxy — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java (hunk 1 above)
Comment (AnonHxy) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java]: Could you please do your final comment @merlimat

**Comment by AnonHxy** [comment]
This patch  broken `org.apache.pulsar.client.impl.BatchMessageContainerImplTest.recoveryAfterOom` test. Please help fix it , thanks @tjiuming

**Comment by tjiuming** [comment]
> This patch broken `org.apache.pulsar.client.impl.BatchMessageContainerImplTest.recoveryAfterOom` test. Please help fix it , thanks @tjiuming

@AnonHxy Yes, I've fixed, PTAL

**Comment by AnonHxy** [comment]
> > This patch broken `org.apache.pulsar.client.impl.BatchMessageContainerImplTest.recoveryAfterOom` test. Please help fix it , thanks @tjiuming
> 
> @AnonHxy Yes, I've fixed, PTAL

Also have some checkstyle failure. Please make sure that CI could be pass in your own fork repo first~  @tjiuming

**Comment by merlimat** [comment]
@tjiuming @Jason918 @AnonHxy 

> Allocate small memory for batch message first, and let it grow(with no memory resizes and memory copies).

How can you say there are no memory resizes or copies? 

I'm still not convinced that this change is not bringing a significant performance regression.

**Comment by AnonHxy** [comment]
Hi @merlimat , `compositeBuffer` will not lead to memory copies until the `componentCount`  reached the `maxNumComponents`, which default value is 16. 

 If we want to avoid any memory, I think we could set the valule of `maxNumComponents`  equals to `batchingMaxMessages`,  which defalut is 1000. WDYT

**Comment by Jason918** [comment]
@AnonHxy Can we do a quick test to validate this? Focus on the case with large number of messages in one batch.

**Comment by AnonHxy** [comment]
> @AnonHxy Can we do a quick test to validate this? Focus on the case with large number of messages in one batch.

Sure

**Comment by merlimat** [comment]
@AnonHxy If we are writing bytes into the composite buffer, it either needs to have space into its own last buffer or to allocate a buffer and copy data into it.

**Comment by tjiuming** [comment]
@merlimat @AnonHxy @Jason918 
For `CompositeByteBuf`, if the number of components in the buffer not exceeds `maxComponent`, it wound not resize/mem_copy. But if the capital is not enough and need a new component, all the components will merge(copy) into a new component.
Here I use the default value of `maxComponent`, which is 16. Because I'm not sure what exact value to set.
But on the whole, the buffer will be allocated lazily, which means fewer wasted memory. To some extent, it achieves the purpose of saving memory.
To reduce mem_copy/resize, maybe we can calculate the certain value of `maxComponent` with `maxBatchSize`, is it required?

**Comment by merlimat** [comment]
> Here I use the default value of maxComponent, which is 16. Because I was not sure what exact value to set.
> But on the whole, the buffer will be allocated lazily, which means fewer wasted memory. To some extent, it achieves the purpose of saving memory.

Allocating many small buffers will have a significant large impact compared to allocating 1 single large buffer: 
 1. Netty allocator synchronization
 2. Scattered write on the socket from many small buffers

**Comment by tjiuming** [comment]
@merlimat 
Yes, there is allocator synchronization, but in `ProducerImpl#sendAsync(Message<?> message, SendCallback callback)`:
   ```
            try {
            synchronized (this) {
              ......
            }
        } catch (PulsarClientException e) {
             ......
        } catch (Throwable t) {
            ......
        }
   ``` 
 which means for a Producer, only 1 thread could send message at the same time, I think even there is allocator synchronization, it wouldn't have a significant large impact.

> Scattered write on the socket from many small buffers

For this part, I don't think it will make a big difference. Yes, read bytes from `CompositeByteBuf` is truly slower than 1 single large buffer, because it need to calculate the `readerIndex`, but when write it to channel, `CompositeByteBuf` will transfer to `ByteBuffer[]`, it would not too much slower than `ByteBuffer`.

**Comment by AnonHxy** [comment]
I have tested this feature using standalone cluster.  It seems that  this PR may bring a performance regression. The reason is that:
When using `ByteBufAllocator#buffer(int)`, the Bytebuf writing took only a little cpu time:
image

And when we using `ByteBufAllocator#compositeBuffer`, the Bytebuf wrte took more cpu time:
image

I think another solution is that we can introduce a way to shrink the `maxBatchSize`. For example,  if current batch size is less that `3/4 * maxBatchSize` we will let `maxBatchSize = 3/4 * maxBatchSize`

@merlimat @tjiuming @Jason918 @codelipenghui

**Comment by Jason918** [comment]
@AnonHxy Is there a significant throughput difference between these two?

**Comment by AnonHxy** [comment]
>  Is there a significant throughput difference between these two?

OK.  @Jason918 
1. First I setup up a standalone cluster using apache-pulsar-2.10.1
2. Then started a producer like below to publish messages, starting with command `~/houxiaoyu/jdk-17.0.4/bin/java -Xms2000m -Xmx2000m -XX:MaxDirectMemorySize=4000m -jar pulsar-buffer.jar`:
```
  public static void main(String[] args) throws Exception {
        String topic = "hxy-test-mem";
        byte[] smallMsg = new byte[64];
        PulsarClient client = PulsarClient.builder()
                .serviceUrl("pulsar://127.0.0.1:6650")
                .memoryLimit(2000, SizeUnit.MEGA_BYTES)  // memory limit 2000M
                .build();
        Producer<byte[]> producer = client.newProducer()
                .topic(topic)
                .enableBatching(true)
                .create();
        for (int i = 0; i < Integer.MAX_VALUE; ++i) {
            producer.sendAsync(smallMsg);
        }
    }
```
3.  I have compared the producer stats three times, and all of them came to the same conclusion that throughput will be descended when using `compositeBuffer`.  Here is the details logs:
   *  first time test:
  
      buffer：
image
     compositeBuffer：
image

  * second time test
  
    buffer:
image
    compositeBuffer：
image

   * third time test
   
    buffer:
image

   compositeBuffer：   
image

**Comment by AnonHxy** [comment]
And here is the attachment of Flame Graph
[buffer.html.zip](https://github.com/apache/pulsar/files/9644855/buffer.html.zip)
[compositeBuffer.html.zip](https://github.com/apache/pulsar/files/9644857/compositeBuffer.html.zip)

**Pull request merged**